content/derived-datatypes.rst

```diff
@@ -11,47 +11,461 @@ Derived datatypes
+         // how many of each type in a "block" of Pair
+         int block_lengths[2] = {1, 1};
+         // displacements of data members in Pair
+         MPI_Aint displacements[2];
+         // why not use pointer arithmetic directly?
+         MPI_Get_address(&my_pair.first, &displacements[0]);
+         MPI_Get_address(&my_pair.second, &displacements[1]);
+
+         // create and commit the new type
+         MPI_Datatype mpi_pair;
+         MPI_Type_create_struct(2, block_lengths, displacements, typesig, &mpi_pair);
+         MPI_Type_commit(&mpi_pair);
+
+      What are the size and the extent? Do they match up with our pen-and-paper calculation?
+      Try different combinations of datatypes and adding other fields to the ``struct``.
+
+      Download a :download:`working solution <code/struct-extent-size-solution.c>`
 
-.. todo::
 
-   - Draw diagram showing how to calculate the extent from a typemap.
-   - Type-along showing how to get extent and size of basis datatypes.
+.. typealong:: Extents and the ``count`` parameter
+
+   Let us reiterate: the extent of a custom datatype *is not* its size. The
+   extent tells the MPI runtime how to get to the **next** item in an array of a
+   given type, much like a *stride*.
+
+   We can send an array of ``n`` ``int``-s with a single |term-MPI_Send|:
+
+   .. literalinclude:: code/snippets/send_n.c
+      :language: c
+      :lines: 33-45
+
+   or with ``n`` such calls:
+
+   .. literalinclude:: code/snippets/n_send.c
+      :language: c
+      :lines: 35-46
+
+   In the latter case, we must program explicitly how to get the next element in
+   the array by using the extent of the datatype.
+
+
+Packing and unpacking
+---------------------
+
+MPI offers the possibility to pack and unpack data of known datatype into a
+single contiguous memory buffer, *without* first having to define a
+corresponding datatype.
+This can be an extremely useful technique to reduce messaging traffic and could
+help with the readability and portability of the code.
+The resulting packed buffer will be of type ``MPI_PACKED`` and can contain any
+sort of heterogeneous collection of basic datatypes recognized by MPI.
+
+
+.. figure:: img/E01-pack_unpack.svg
+   :align: center
+
+   MPI allows the programmer to communicate heterogeneous collections into a
+   single message, without defining a full-fledged custom datatype. The data is
+   packed into a buffer of type ``MPI_PACKED``. On the receiving end, the buffer
+   will be unpacked into its constituent components.
+
+
+.. signature:: |term-MPI_Pack|
+
+   Pack data in noncontiguous memory to a contiguous memory buffer.
+
+   .. code-block:: c
+
+      int MPI_Pack(const void *inbuf,
+                   int incount,
+                   MPI_Datatype datatype,
+                   void *outbuf,
+                   int outsize,
+                   int *position,
+                   MPI_Comm comm)
+
+
+.. figure:: img/E01-pack.svg
```

Comment:
I don't understand what the pink color represents (or indeed the whoel graphic). Perhaps separating input and output depictions would help?

Reply:
Pink is supposed to be data already packed in `outbuf`. Looking at the picture again, I agree it's confusing. I can reword the caption.

Reply:
Figures remade and with a (hopefully) better worded caption!